### PR TITLE
#333: aggregated reports

### PR DESCRIPTION
**Pull request by MihaStemberger**
Hello, here is my shot at trying to solve https://github.com/mojohaus/versions-maven-plugin/issues/333

### CODE REVIEW

File: versions-common/src/main/java/org/codehaus/mojo/versions/utils/PropertyComparator.java [hunk 1]
```diff
@@ -42,14 +41,18 @@ public enum PropertyComparator implements Comparator<Property>
...
                 : p1 == null
-                    ? 1
-                    : p2 == null
-                        ? -1
-                        : StringUtils.compare( p1.getName(), p2.getName() );
+                        ? 1
+                        : p2 == null
+                                ? -1
+                                : ( r = StringUtils.compare( p1.getName(), p2.getName() ) ) == 0
+                                        ? StringUtils.compare( p1.getValue(), p2.getValue() )
```
Comment (ajarmoniuk) [inline versions-common/src/main/java/org/codehaus/mojo/versions/utils/PropertyComparator.java:55]: The original idea behind omitting the value in comparison was that properties with different values were treated as the same property. Was there a reason you wanted to treat properties with different values as different properties?

File: versions-common/src/main/java/org/codehaus/mojo/versions/api/Property.java [hunk 2]
```diff
@@ -176,4 +184,14 @@ public void setBanSnapshots( boolean banSnapshots )
     {
         this.banSnapshots = banSnapshots;
     }
+
+    public String getValue()
```
Comment (ajarmoniuk) [inline versions-common/src/main/java/org/codehaus/mojo/versions/api/Property.java:188]: The original idea behind omitting the value in comparison was that properties with different values were treated as the same property. That's why the Property class didn't have a value property.

**Review by ajarmoniuk — CHANGES_REQUESTED** [review]

File: versions-maven-plugin/src/it/it-dependency-updates-aggregate-report-issue-333/src/site/site.xml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<project>
+  <skin>
+    <groupId>org.apache.maven.skins</groupId>
+    <artifactId>maven-default-skin</artifactId>
+    <version>1.0</version>
+  </skin>
+</project>
```
[the remark this answers is not in the document]
Comment (MihaStemberger) [reply versions-maven-plugin/src/it/it-dependency-updates-aggregate-report-issue-333/src/site/site.xml]: Added

File: versions-common/src/main/java/org/codehaus/mojo/versions/utils/PropertyComparator.java (hunk 1 above)
Comment (MihaStemberger) [reply versions-common/src/main/java/org/codehaus/mojo/versions/utils/PropertyComparator.java:55]: The reason is that a child may define its own property value. If comparison is done only by name, then only one property entry will be in the final report and not all, that are possible.

File: versions-common/src/main/java/org/codehaus/mojo/versions/api/Property.java (hunk 2 above)
Comment (MihaStemberger) [reply versions-common/src/main/java/org/codehaus/mojo/versions/api/Property.java:188]: [See above](#discussion_r1038304268-permalink)

File: versions-maven-plugin/src/it-repo/dummy-aggregate-child-333.pom [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  </parent>
+
+  <groupId>localhost</groupId>
+  <artifactId>dummy-aggregate-child</artifactId>
+  <version>1.0.1-SNAPSHOT</version>
+  <packaging>pom</packaging>
+
+  <properties>
+    <version.maven-clean-plugin>2.1</version.maven-clean-plugin>
+    <version.maven-deploy-plugin>2.1</version.maven-deploy-plugin>
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/it-repo/dummy-aggregate-child-333.pom:18]: Can we use a little newer versions in test or a dummy dependency

**Comment by slawekjaranowski** [comment]
For new files we use standard Apache 2.0 license headers: 
https://groups.google.com/g/mojohaus-dev/c/OTg7XfiX61w

I prefer groovy scripts for verification than beanshell ... as nice to have 😄

File: versions-maven-plugin/src/it/it-plugin-updates-aggregate-report-issue-333/pom.xml [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  <reporting>
+    <plugins>
+      <plugin>
+        <artifactId>maven-project-info-reports-plugin</artifactId>
+        <version>2.1</version>
+        <reportSets>
+          <reportSet/>
+        </reportSets>
+      </plugin>
```
[the remark this answers is not in the document]
Comment (MihaStemberger) [reply versions-maven-plugin/src/it/it-plugin-updates-aggregate-report-issue-333/pom.xml]: Added the configuration to `dummy-aggregate-parent-333.pom` and removed `maven-project-info-reports-plugin` from all child poms

File: versions-maven-plugin/src/it-repo/dummy-aggregate-child-333.pom (hunk 4 above)
Comment (MihaStemberger) [reply versions-maven-plugin/src/it-repo/dummy-aggregate-child-333.pom]: Using newer version of `maven-clean-plugin`, also revisited and changed test poms, so they mainly use dummy dependencies.

**Comment by MihaStemberger** [comment]
> For new files we use standard Apache 2.0 license headers: https://groups.google.com/g/mojohaus-dev/c/OTg7XfiX61w
> 
> I prefer groovy scripts for verification than beanshell ... as nice to have smile

Added `Apache 2.0 license headers` to newly added *.java files, replaced beanshell verification scripts with groovy scripts

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractDependencyUpdatesReportMojo.java [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+import org.codehaus.mojo.versions.reporting.model.DependencyUpdatesModel;
+import org.codehaus.mojo.versions.utils.DependencyComparator;
+import org.codehaus.mojo.versions.xml.DependencyUpdatesXmlReportRenderer;
+import org.codehaus.plexus.i18n.I18N;
+
+/**
+ * Generates a report of available updates for the dependencies of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractDependencyUpdatesReportMojo.java:56]: copy paste 😄

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPluginUpdatesReportMojo.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+import org.codehaus.mojo.versions.reporting.model.PluginUpdatesModel;
+import org.codehaus.mojo.versions.utils.PluginComparator;
+import org.codehaus.mojo.versions.xml.PluginUpdatesXmlReportRenderer;
+import org.codehaus.plexus.i18n.I18N;
+
+/**
+ * Generates a report of available updates for the plugins of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPluginUpdatesReportMojo.java:49]: copy paste 😄

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPropertyUpdatesReportMojo.java [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+import org.codehaus.mojo.versions.utils.PropertyComparator;
+import org.codehaus.mojo.versions.xml.PropertyUpdatesXmlReportRenderer;
+import org.codehaus.plexus.i18n.I18N;
+
+/**
+ * Generates a report of available updates for properties of a project which are linked to the dependencies and/or
+ * plugins of a project.
+ *
+ * @author Stephen Connolly
+ * @since 1.0-beta-1
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/AbstractPropertyUpdatesReportMojo.java:48]: copy paste 😄

File: versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/PluginUpdatesAggregateReportMojo.java [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                                                org.eclipse.aether.RepositorySystem aetherRepositorySystem,
+                                                Map<String, Wagon> wagonMap,
+                                                ReportRendererFactory rendererFactory )
+    {
+        super( i18n, repositorySystem, aetherRepositorySystem, wagonMap, rendererFactory );
+    }
+
+    /**
+     * {@inheritDoc}
+     * */
```
Comment (slawekjaranowski) [inline versions-maven-plugin/src/main/java/org/codehaus/mojo/versions/PluginUpdatesAggregateReportMojo.java:55]: not need IMHO

**Review by slawekjaranowski — COMMENTED** [review]
last nits

**Comment by MihaStemberger** [comment]
> last nits

Resolved

**Comment by ajarmoniuk** [comment]
Yes, you're supposed to have edited https://github.com/mojohaus/versions-maven-plugin/blob/master/versions-maven-plugin/src/site/markdown/index.md#reporting-goals-overview 

But I guess you can still do it in another PR ;)

**Pull request merged**